### ISSUE TITLE
Ideas for a refresh of the Gambit website

### ISSUE THREAD

**Issue by lassik** [gambit/gambit#592]
(Continuing from #575)

@daviwil did fantastic work on a draft based on some copy I wrote last year.

**Comment by lassik** [comment]
It might be good to have a comparison on the front page between Gambit and other  language implementations that people interested in it might be looking at. In particular (please correct any misunderstandings I may have):

* Since Gambit uses Scheme, it is simpler than Common Lisp implementations like SBCL. On the other hand, image-based development is generally easier with Common Lisp than Scheme.

* Gambit has a much smaller runtime than JVM/CLR implementations like Clojure so it starts much faster and can be deployed in many more scenarios.

* Since Gambit compiles to C, it is more portable than just about anything and is more flexible in what can be left out of the resulting binary.

* Gambit can do many of the same things as Erlang but is faster, lighter weight, more flexible and more portable. Scheme and Erlang are about equally weird :) Erlang currently has more battle-tested hot-loading of modules than Gambit does.

* Gambit can do many of the same things as Go (concurrency, speed, networking, OS/FFI) while offering a much higher level of abstraction thanks to Scheme.

* Unlike Go, Erlang, Python, Ruby, Clojure - Scheme is a standard language.

**Comment by lassik** [comment]
* A rough performance comparison of Gambit and OCaml and MLton would be very interesting. Thanks to Scheme, Gambit is at a higher level of abstraction and is easier for exploratory programming. ML is probably a bit faster and can catch more errors at compile time. Scheme has a simpler and much more flexible syntax.

* A comparison to Chez Scheme is probably in order. A major difference is that Chez is an incremental compiler (similar to most Common Lisp implementations) whereas Gambit is split between a batch compiler and a separate interpreter. This means that Gambit is more portable and possibly easier to embed to applications, whereas Chez offers a more seamlessly integrated REPL experience. (Not quite confident I got everything right here.)

**Comment by lassik** [comment]
* The obligatory performance comparison to LuaJIT would be fun to have. Unlike Lua, Scheme is standardized and is a more fully fleshed out language. Scheme's syntax is more uniform and the REPL is probably more useful. What kind of debugger does LuaJIT have? Lua's runtime system is way smaller than Gambit's; what about LuaJIT's? Gambit is portable to as many platforms as Lua, and to more platforms than the LuaJIT native compiler.

**Comment by lassik** [comment]
To match the **About Gambit** section of the front page, here's a similar blurb for Scheme:

## About Scheme

Scheme is a classic programming language in the Lisp family. It is known among Lisp dialects as clean and minimalist and is a favorite teaching language. Scheme encourages an impure functional coding style using recursion instead of iteration, with closures to keep state.

Unlike most functional languages, Scheme uses dynamic typing. This makes the language simpler, easier to learn and much more flexible than static typing. The tradeoff is that fewer errors are caught at compile time. Thanks to dynamic typing and homoiconic syntax, Scheme code is just data. This enables the most advanced macro system in any language.

**Comment by feeley** [comment]
I like all your points.  However there might be a TMI problem to put all of that on the landing page that has to be mostly "to the point", i.e. high-level selling points and quick installation instructions.

**Comment by lassik** [comment]
True, the comparison would probably go better on a separate page.

**Comment by daviwil** [comment]
Yep, I agree, I think we can gather more in-depth "should I use this?" information on another page.  And maybe it's just me, but I don't want to necessarily point out reasons why someone _shouldn't_ want to use Gambit or Scheme ;)  All of the information you mentioned will certainly be useful to convince folks to try Gambit though!

**Comment by lassik** [comment]
> I don't want to necessarily point out reasons why someone _shouldn't_ want to use Gambit or Scheme ;)

That depends a lot on which audience it is aimed at. Gambit is a research system so Marc has written lots of academic papers and published benchmarks where an even-handed comparison is expected. Gambit will also appeal more to an audience of "hardcore" or experienced programmers than something geared more toward convenience like Ruby, and in any case performance for complex applications is a big selling point. In complex evaluations like that people probably appreciate a more thorough and fact-based view.

We also talked about Marc last year that the reason Gambit is possibly hard to sell (in terms of expanding mindshare) because it sounds like it has too many good features in one system. I think many people instinctively disregard anything that seems too good to be true and don't take a closer look at it. To that end, a detailed exposition provides background information to justify why it has so many capabilities. In Gambit's case one clear reason is that it has been developed for 30 years. Providing a nuanced view of performance and capabilities will also show that there are individual tasks where something else is better, which gives a sense of realism (as opposed to mass marketing with a "best for every job" vibe). Admitting weaknesses gives more credibility to claims of strength; that credibility should be an asset when there seem to be too many strengths.

A good point of comparison might be OpenBSD, which is also a highly esteemed project with an expert userbase. Their reputation is very good even though their marketing is down-to-earth and detailed and features no outlandish claims.

**Comment by lassik** [comment]
Then again, Rust is also a favorite of the high-performance/expert contingent and their marketing doesn't seem particularly modest and balanced. I've never heard them say that "borrowing is great and all, but sometimes you would really be better off with GC" :) I wonder what part Mozilla's backing has played in giving it a positive spin.

And Ruby is great for novices, but their marketing has always been strikingly modest instead of brash (at least in Europe/Asia). These things can be quite idiosyncratic.

**Comment by daviwil** [comment]
One thing I've noticed (even in myself) is that developers might tend to fixate on one seemingly negative detail to convince themselves to not use something even when everything else is great.  I think the detail for me in the description about Scheme above is "The tradeoff is that fewer errors are caught at compile time."  That's certainly true, but it sounds a little scary.

I'm all for even-handed analysis though, I think it fits with Gambit's ethos.  I'd be fine even mentioning the detail I discussed above because I don't think it will matter a whole lot in the long run.  The people who want to use Scheme will pick Gambit because of its speed and capabilities (which we should discuss and compare at length).  Those details are what brought me to it!

**Comment by lassik** [comment]
> One thing I've noticed (even in myself) is that developers might tend to fixate on one seemingly negative detail to convince themselves to not use something even when everything else is great. I think the detail for me in the description about Scheme above is "The tradeoff is that fewer errors are caught at compile time." That's certainly true, but it sounds a little scary.

That's a very good point. A corollary is that programmers can believe one feature magically solves all their problems.

This comes up with dynamic vs static typing. Static typing enthusiasts often seem to effectively say that static typing produces working software and dynamic typing produces buggy software (exaggerating only a little). I've never understood how such a nuanced issue can be seen in such simplistic terms but apparently many advocates do. In that respect, admitting that "fewer errors are caught at compile time" could be twisted to mean that "Scheme programs are buggy and switching to ML or Haskell will solve all your programming problems".

**Comment by lassik** [comment]
Rust's borrow checker seems to elicit a similar response as static typing more generally, perhaps in many of the same people. It's worth trying to understand how that phenomenon works psychologically. It seems some people are predisposed to think safety is a goal that can be clearly defined and directly pursued, and are elated whenever new safety features are added. Speed is another similar concept - lots of people are impressed whenever anything "is faster" by some metric that apparently clearly approximates real-world performance.

Lisp and Scheme appeal to people who think safety and performance are inherently ill-defined or situation-dependent concepts and hence see them as merely two factors among many. But that doesn't sound as good in a brochure as saying it is fast and safe :)
